### PR TITLE
8264795: IGV: Upgrade NetBeans platform

### PR DESCRIPTION
**Pull request by robcasloz**
This change upgrades the NetBeans platform on which IGV is based to its latest version ([12.3](https://netbeans.apache.org/download/nb123/index.html)) and switches IGV's build system from Ant to Maven. The upgrade introduces support for a wide range of JDK versions (from 8 to 15, the latest version supported by NetBeans 12.3), and the switch from Ant to Maven makes the IGV build simpler, faster (first-time build is approximately 5x faster), and more stable (all dependencies are fetched directly from the Maven central repository).

The change also fixes broken unit tests in the Data module and runs them by default when building.

#### Testing

Regression-tested the following use cases manually on all combinations of (Linux, Windows, macOS) and (JDK 8, JDK 11, JDK 15):

- build
- open graph file (small.xml in [test-input.zip](https://bugs.openjdk.java.net/secure/attachment/93988/test-input.zip))
- import graphs via network (localhost)
- expand groups in outline
- open a graph
- open a clone
- zoom in and out
- search a node
- apply filters
- extract a node
- show all nodes
- select nodes corresponding to a bytecode
- view control flow
- select nodes corresponding to a basic block
- cluster nodes
- show satellite view
- compute the difference of two graphs
- change node text
- remove a group
- save groups into a file
- open a large graph file (large.xml in [test-input.zip](https://bugs.openjdk.java.net/secure/attachment/93988/test-input.zip))
- open a large graph ("After Escape Analysis" in large.xml)

Thanks to Vladimir Ivanov for helping with testing on macOS.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264795](https://bugs.openjdk.java.net/browse/JDK-8264795): IGV: Upgrade NetBeans platform

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 11fb43faeee11d99aa3c13240c4e340f7116f687
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 11fb43faeee11d99aa3c13240c4e340f7116f687
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - no project role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3361/head:pull/3361` \
`$ git checkout pull/3361`

Update a local copy of the PR: \
`$ git checkout pull/3361` \
`$ git pull https://git.openjdk.java.net/jdk pull/3361/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3361`

View PR using the GUI difftool: \
`$ git pr show -t 3361`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3361.diff">https://git.openjdk.java.net/jdk/pull/3361.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Great work!
After this is pushed consider to update wiki page too: https://wiki.openjdk.java.net/display/HotSpot/IdealGraphVisualizer

**Comment by robcasloz** [comment]
> Great work!
> After this is pushed consider to update wiki page too: https://wiki.openjdk.java.net/display/HotSpot/IdealGraphVisualizer

Thanks for reviewing, Vladimir! Yes, I plan to do that.

File: src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/serialization/Parser.java [hunk 1]
```diff
@@ -145,12 +146,17 @@ protected Group start() throws SAXException {
...
-                    public void run() {
-                        parent.addElement(group);
-                    }
-                });
+                Runnable addToParent = new Runnable(){
+                        @Override
+                        public void run() {
+                            parent.addElement(group);
+                        }
+                    };
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/serialization/Parser.java:154]: You could directly use `Runnable addToParent = () -> parent.addElement(group);` and also for the other ones below.

**Review by chhagedorn — APPROVED** [review]
Very nice work!

**Review by neliasso — APPROVED** [review]
I have tested IGV on a High-DPI screen (4K). The screens are attached to the bug-report.

Compared to baseline the new version improved DPI scaling on both JDK8 and JDK15. Some elements still doesn't scale.

Remaining issues identified:
1) The default zoom level is to small
2) The filter view has broken scaling - the font scales but not the line height.
3) The buttons above the graph view and file list doesn't scale - they are still tiny.
4) The splash screen doesn't scale

Still this is a big improvement and the remaining issues can be solved in separate PRs.

Looks good - Reviewed!

File: src/utils/IdealGraphVisualizer/README.md [hunk 2]
```diff
@@ -3,22 +3,17 @@
...
+included as part of HotSpot since that was the primary target of the tool. The
 tool itself is fairly general with only a few modules that contain C2 specific
 elements.
 
-The tool is built on top of the NetBeans 7 rich client infrastructure and so
-requires NetBeans to build. It currently requires at least Java 6 to run as it
-needs support for JavaScript for its filtering mechanism and assumes it's built
-into the platform.  It should build out of the box with NetBeans 7.0 and Java 6
-or later.
+The tool is built on top of the NetBeans Platform, and requires Java 8 or later.
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/utils/IdealGraphVisualizer/README.md]: Thanks, done!

File: src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/serialization/Parser.java (hunk 1 above)
Comment (robcasloz) [reply src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/serialization/Parser.java]: Good suggestion, thanks, done!

**Comment by robcasloz** [comment]
> Very nice work!

Thanks for reviewing, Christian!

**Comment by robcasloz** [comment]
> I have tested IGV on a High-DPI screen (4K). The screens are attached to the bug-report.
> 
> Compared to baseline the new version improved DPI scaling on both JDK8 and JDK15. Some elements still doesn't scale.
> 
> Remaining issues identified:
> 
>     1. The default zoom level is to small
> 
>     2. The filter view has broken scaling - the font scales but not the line height.
> 
>     3. The buttons above the graph view and file list doesn't scale - they are still tiny.
> 
>     4. The splash screen doesn't scale
> 
> 
> Still this is a big improvement and the remaining issues can be solved in separate PRs.
> 
> Looks good - Reviewed!

Thanks for reviewing, Nils! I will create a separate RFE for DPI scaling issues.

I also added a clarification to the README file (https://github.com/openjdk/jdk/pull/3361/commits/e84d171e12a238cf0a1905a4bc8107ac6141a540) based on our offline discussion.

**Review by chhagedorn — APPROVED** [review]

**Review by navyxliu — APPROVED** [review]
Thank you for modernizing IGV. I manage to import this project to IntelliJ. The process is hassle-free!

I also ran using jdk8/jdk11 on macOS. I haven't identified any problem so far.

**Comment by kelthuzadx** [comment]
Nice work. It works on Windows home pc.

**Comment by robcasloz** [comment]
> Thank you for modernizing IGV. I manage to import this project to IntelliJ. The process is hassle-free!
> 
> I also ran using jdk8/jdk11 on macOS. I haven't identified any problem so far.

Thanks for checking, Xin!

File: src/utils/IdealGraphVisualizer/Filter/pom.xml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        <dependency>
+            <groupId>org.netbeans.api</groupId>
+            <artifactId>org-openide-filesystems</artifactId>
+            <version>${netbeans.version}</version>
+        </dependency>
+    </dependencies>
+    <profiles>
+        <profile>
+            <activation>
+                <jdk>[15,)</jdk>
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/utils/IdealGraphVisualizer/Filter/pom.xml:93]: > Does this mean that module Filter will depend on external nashorn after jdk15?

Yes (including JDK 15).

> I know that jdk doesn't remove nashorn util jdk15, but does it make sense that IGV always picks up external nashorn. My concern is "javascript" in jdk8/11 may be not well-maintained and leave vulnerability behind.

I assume potential issues in "built-in" Nashorn are addressed in JDK 8-14 as for any other OpenJDK component. Besides, [standalone (external) Nashorn only works on JDK >= 11](https://github.com/szegedi/nashorn/wiki/Using-Nashorn-with-different-Java-versions), and I think it makes sense that IGV stills support JDK 8 for minimal disruption (this was the only JDK version supported before this upgrade).

**Comment by robcasloz** [comment]
> Nice work. It works on Windows home pc. rocket

Thanks for checking, Yang!

**Comment by robcasloz** [comment]
/summary
Upgrade IGV's underlying NetBeans platform to version 12.3, switch build system
from Ant to Maven, and fix broken unit tests in Data module.

**Pull request closed without merging**